CityOfDetroit/COD-Design-System
# Mixins for handling observed attributes & CSS classes

## Fixes #127

## Context

As discussed in depth in #134 - [specifically this comment](https://github.com/CityOfDetroit/COD-Design-System/issues/134#issuecomment-1871446638) - we cannot rely on custom element constructors being called in the order in which they are parsed in the DOM tree. It's possible custom elements are upgraded in an order that is different than the DOM tree, and the constructors would follow suit.

The consequence of this observation is that any time we want to pass a CSS class down the DOM tree of custom elements, we must do so with [observed attributes](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements#responding_to_attribute_changes) as well as `slotchange` events.

## This PR

1. Add a [mixin](https://en.wikipedia.org/wiki/Mixin) to handle observed "class attribute" changes. Here we're defining a "class attribute" as an attribute that controls the inclusion of a class on an HTMLElement, since this seems to be a common pattern on our web components. This "class attribute" pattern is used often because [Shadow DOM Encapsulation](https://javascript.info/shadow-dom#encapsulation) prevents us from using complex CSS selector descendant combinators, so the class must be explicitly passed down from the parent element.
2. Add a mixin to handle general observed attribute changes.
3. Add a utility to propagate stacked table attributes to child elements used during `slotchange` events.
4. Replace previous attribute and class propagation implementation with these new mixins and utility for the `<cod-table>` component.
5. Update the import order of components in `index.js` to import them in the opposite order they're using in the DOM (since import order shouldn't matter).

## Alternatives

I laid out a couple alternatives in CityOfDetroit/COD-Design-System#127 but they were non-starters.

A couple other alternatives that aren't necessarily non-starters:

1. Continue writing one-off implementations for observed attribute change handlers. This approach might be fine if the observed attribute has some unique handler behavior, but as we see in this PR, there are a couple main types of observed attributes ("class attributes" being one of those) that share a common pattern of implementation.
2. Use [custom states and custom state pseudo-class CSS selectors](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements#custom_states_and_custom_state_pseudo-class_css_selectors) for managing the 'stacked' styles. Semantically, custom states represent state that can be modified with user interaction. It feels like a bit of a stretch in this case since user interaction isn't going to change whether the table is stacked. Also, custom internal state doesn't solve the problem of propagating that state down the DOM tree where it's needed for styling.

## Testing

1. Build the `main.js` bundle using Webpack: `yarn build`
2. Use this branch as a composer package ([see wiki](https://intranet.detroitmi.gov/node/2926)) and then test the changes on a Drupal test site with the package on this branch in use:

![test-table](https://github.com/CityOfDetroit/COD-Design-System/assets/143553259/b51a59f1-747c-4fa8-9543-f8068d899cc4)
![test-table-mobile](https://github.com/CityOfDetroit/COD-Design-System/assets/143553259/9fb81fb2-ac6c-4011-8472-630d406bc037)
